### PR TITLE
Replace idle sleep with a wakeable SDL wait

### LINKED ISSUE #283: [Feature]: Replace idle sleep with a wakeable event wait to reduce typing latency
**Title:** [Feature]: Replace idle sleep with a wakeable event wait to reduce typing latency

**Labels:** feature, medium

## Status Quo

Architect's main loop polls SDL events, processes PTY output, renders if needed, and then sleeps when it considers the app idle. The idle sleep target is currently 50 ms in `src/app/runtime.zig`. Keyboard input is written to the PTY immediately, but if a key arrives just after the loop enters idle sleep, the app does not wake until that sleep finishes. That is enough to make typing feel slower than Ghostty even when the rest of the input path is reasonably efficient.

## Objectives

Reduce perceived typing latency so Architect feels responsive during normal terminal use. The app should still conserve power when idle, but it should wake promptly for keyboard, mouse, and other user-visible events instead of waiting for a fixed sleep interval to expire.

## User Flow

**Trigger:** The user types into a focused terminal after a short idle period.

1. Architect waits in an idle state without burning CPU.
2. A keyboard or other wake-worthy event arrives.
3. The frame loop wakes immediately, handles the input, drains terminal output, and renders the updated frame.

**Result:** Typing latency no longer includes an avoidable idle-sleep delay.

## Scope

**In scope:**

- Replace the fixed idle sleep with a wakeable wait strategy in the main loop.
- Preserve current active and idle pacing for animations and normal rendering.

**Out of scope:**

- Changing terminal emulation behavior or shell echo semantics.
- Adding speculative local echo before PTY output returns.

### Affected Modules

- `src/app/runtime.zig`: replace idle sleep with wakeable waiting and update frame pacing logic.
- `src/session/notify.zig`: make sure external notifications are still observed promptly by the runtime loop.
- `src/platform/sdl.zig`: add any small SDL helper needed for waiting behavior.

### Tasks

1. [ ] Define the wakeable idle strategy and the conditions that should break idle waiting. `src/app/runtime.zig`
2. [ ] Replace fixed idle sleeping with an SDL-backed wait path that still respects animation and render deadlines. `src/app/runtime.zig`
3. [ ] Verify notification delivery, quit flow, and input handling still behave correctly with the new wait model. `src/app/runtime.zig`, `src/session/notify.zig`
4. [ ] Write tests for frame-pacing helper logic, wake and deadline calculations, and idle versus active transitions.
5. [ ] Update `docs/ARCHITECTURE.md` to describe the new frame-loop waiting behavior.

### New Dependencies

None

## Acceptance Criteria

- [ ] Idle typing no longer waits on a fixed 50 ms sleep window.
- [ ] Active rendering and animations still run at the intended cadence.
- [ ] External notifications and quit handling remain responsive.
- [ ] Tests cover the happy path and documented edge cases.
- [ ] `docs/ARCHITECTURE.md` is updated.
- [ ] Pre-commit hooks pass.

### PR DESCRIPTION
**Pull request by forketyfork**
Fixes #283

## Summary

Architect used a fixed 50 ms idle sleep in the main loop, which meant a keypress or other visible event could land just after the loop went idle and then wait for that sleep to finish. This changes the frame loop to compute an explicit wait strategy, use `SDL_WaitEventTimeout(...)` while idle, keep the existing active pacing rules, and let the notification thread wake the runtime immediately by posting a custom SDL event after it queues external updates.

## Testing

- [x] `zig build`
- [x] `zig build test`
- [x] `just lint`

## Test plan

- [ ] Leave Architect idle for a few seconds, then type in a focused terminal and confirm input appears immediately instead of feeling delayed by the old idle sleep window.
- [ ] Trigger an external session status notification while Architect is idle and confirm the status highlight updates right away.

### CODE REVIEW

File: src/app/runtime.zig [hunk 1]
```diff
@@ -93,6 +99,41 @@ fn countSpawnedSessions(sessions: []const *SessionState) usize {
...
 
+fn remainingFrameBudgetNs(target_frame_ns: i128, frame_ns: i128) u64 {
+    if (frame_ns >= target_frame_ns) return 0;
+    return @intCast(target_frame_ns - frame_ns);
+}
+
+fn waitTimeoutMsFromNs(remaining_ns: u64) c_int {
+    const timeout_ms = remaining_ns / std.time.ns_per_ms;
+    if (timeout_ms == 0) return 0;
+
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/app/runtime.zig]: `waitTimeoutMsFromNs` now rounds up to the next whole millisecond and still clamps to `c_int` max. That keeps idle pacing from waking early while input and notifications can still break the wait immediately.

File: src/platform/sdl.zig [hunk 2]
```diff
@@ -119,6 +127,29 @@ pub fn init(
...
+
+    var event = std.mem.zeroes(c.SDL_Event);
+    if (!c.SDL_WaitEventTimeout(&event, timeout_ms)) return null;
+    return event;
+}
+
+pub fn pushWakeEvent(platform_handle: *const Platform) void {
+    var event = std.mem.zeroes(c.SDL_Event);
+    event.user.type = platform_handle.wake_event_type;
+    _ = c.SDL_PushEvent(&event);
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/platform/sdl.zig]: `pushWakeEvent` now checks the result of `SDL_PushEvent` and logs `SDL_GetError()` when SDL rejects the wake event. That keeps the failure mode visible instead of silently falling back to the idle timeout.

File: src/session/notify.zig [hunk 3]
```diff
@@ -240,3 +265,33 @@ test "NotificationQueue - push and drain" {
...
+
+    var wake_count: usize = 0;
+    const wake = RuntimeWake{
+        .context = &wake_count,
+        .callback = TestWake.onWake,
+    };
+
+    try queue.push(allocator, .{ .status = .{ .session = 7, .state = .done } });
+    wake.notify();
+
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/session/notify.zig]: The queue-and-wake path now goes through a shared `enqueueNotification` helper, and the tests exercise that helper directly. They cover both the successful queue-then-wake case and the queue-failure case where no wake is sent.
